### ISSUE TITLE
Abstract Data Definitions

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1166]
We need a better way to describe data definitions that are defined by some abstract function that will be specified at the implementation stage. Currently we use a hack where, in cases where we need to display a symbol for an implementation-stage function, we use `asExpr'` and/or `asVC'` to create a stage-independent version of that function.
https://github.com/JacquesCarette/Drasil/blob/193a0337512b1da40045b52c6a46d8a6f28f8e33/code/drasil-code/Language/Drasil/CodeSpec.hs#L216-L224

See the origin of the need for this in #1135 and #1139.

**Comment by peter-michalski** [comment]
`asExpr'` is no longer in the code base, and was replaced by `funcUID`' in Drasil/code/drasil-code/lib/Language/Drasil/CodeSpec.hs :

```
-- FIXME: hack. 
-- | Used for implementation-stage functions that need to be displayed in the SRS.
funcUID' :: Func -> UID
funcUID' f = asVC' f ^. uid

-- FIXME: Part of above hack
-- | Helper for 'funcUID''.
asVC' :: Func -> QuantityDict
asVC' (FDef (FuncDef n _ _ _ _ _)) = vc n (nounPhraseSP n) (Variable n) Real
asVC' (FDef (CtorDef n _ _ _ _))   = vc n (nounPhraseSP n) (Variable n) Real
asVC' (FData (FuncData n _ _))     = vc n (nounPhraseSP n) (Variable n) Real 
```

`asVC'` is referenced in:

1.  Drasil/code/drasil-example/glassbr/lib/Drasil/GlassBR/Symbols.hs 
```
symbolsForTable :: [QuantityDict]
symbolsForTable = inputs ++ outputs ++ tmSymbols ++ map qw specParamVals ++ 
  map qw symbolsWithDefns ++ map qw symbols ++ map qw unitless ++
  map qw inputDataConstraints ++ map asVC' [interpY, interpZ]
```

2.   Drasil/code/drasil-example/glassbr/lib/Drasil/GlassBR/DataDefs.hs 
```
strDisFacEq :: Expr
-- strDisFacEq = apply (sy stressDistFac)
--   [sy dimlessLoad, sy aspectRatio]
strDisFacEq = apply (asVC' interpZ) [str "SDF.txt", sy aspectRatio, sy dimlessLoad]
```

**Comment by JacquesCarette** [comment]
It is unsurprising that both uses are in GlassBR. 

Summary: we need to be able to 'specify', in the spec, an abstract function (i.e. with no implementation) that will be refined at implementation-time. So we're not going to specify the operational details in the requirements.

In some sense, I don't really know **why** we can't do that right now, without the hack. It feels like maybe what's in GlassBR right now is 'backwards' and the `QuantityDict` that `asVC'` is creating should actually have been created in the spec itself. In other words, I suspect that the hack is created by a backwards implementation!  i.e. when we took the code and 'lifted' it to Drasil, we didn't quite do the right thing wrt the interpolation functions.

**Comment by balacij** [comment]
@peter-michalski and I had a quick meeting regarding this. A first-pass at this might look approximately like:

1. Create an "Abstract" QuantityDict (e.g., `newtype AbstractQuantityDict = AQD { _qd :: QuantityDict }`).
   This is a "fake" QuantityDict, where the type indicates that it's one that _will not be implemented 
   equationally_, but implemented through "Choices", and have it registered in the ChunkDB. Depending on how we want to handle them, this implementation for them might be slightly different, but the idea should remain the same.
2. Create a new field in the Choices record for a `Map UID UID` (representing a mapping of  "AbstractQuantityDict"s to "Code" through their UIDs)
3. In the `codeSpec` function (where we generate the "CodeSpec") from an SI, Choices, and [Mod], we would sub out the 
   UIDs of the "AbstractQuantityDicts" with the UIDs of the "Code"-related counterparts.
  a. Error out if any registered AbstractQuantityDict in the ChunkDB doesn't have a counterpart registered in the Choices
     (when generating code).
4. Create "abstract" variants of "interpY" and "interpZ" (the only two variables for which we use "asVC'" for), register their UIDs for appropriate replacement with interpY/interpZ, respectfully, through Choices, and finally remove "asVC'".

Do you have any thoughts on this @JacquesCarette ?

**Comment by JacquesCarette** [comment]
Why can't it be a `QuantityDict`? That has no definition field. I don't understand the extra level of complication.  Also, where to we assume that things are implemented equationally?

Yes, we do need to associate the current `interpY` with this chunk, and this is currently done "backwards" with `asVC'`. But doesn't that mean that we don't need to complicate things (as in the above) by just keeping the rest of the control flow as is?

I think I need to see a deeper analysis of what's going on currently (@peter-michalski 's analysis above was perfect for understanding the issue, but we need more of the flow to understand and evaluate potential solutions.)

**Comment by balacij** [comment]
I'm going to split this up into pieces/comments.

The problem, as Brooks wrote in the original issue post, is that there are functions and data that is manually defined at the implementation level which are lacking appropriately abstracted placeholders. To my understanding, this means that there is a _choice_ that was implemented for a data, but that is not passed through the appropriate channel (e.g., `Choices`).

As @peter-michalski mentioned, this only affects GlassBR. GlassBR has 2 manually written modules (defined in glassbr's ModuleDefs.hs file) that contain manually written "helper" functions for reading in a data table and performing some sort of interpolation. It would appear that they are the "choices" here that are lacking an appropriate placeholder.

Currently, to allow these manually written functions to "enter" the GlassBR program, they were forcibly "lifted" into it through `asVC'`.

**Comment by balacij** [comment]
Ok, now let's see how `asVC'` works:
https://github.com/JacquesCarette/Drasil/blob/5dc10b54aefd5c6ddcef0ce15706b95760f6d7ca/code/drasil-code/lib/Language/Drasil/CodeSpec.hs#L135-L145

`asVC'` takes a `Func` (arguably a lower-level encoding) and recreates it's relevant information as a QuantityDict with a symbol defined at the "Implementation" and "Equational" level. The result is then used for references to this `Func`. In a sense, we are pushing this choice upwards. Then, when code is generated, the UID of the QuantityDict lines up nicely with this `Func`, and so no issues arise. A fairly neat "hack"!

**Comment by balacij** [comment]
`CodeSpec.hs` also contains a similarly named, self-identified "hack", called "asVC":
https://github.com/JacquesCarette/Drasil/blob/5dc10b54aefd5c6ddcef0ce15706b95760f6d7ca/code/drasil-code/lib/Language/Drasil/CodeSpec.hs#L124-L145

They produce nearly identical QuantityDicts through different functions.

`asVC` uses `implVar`, which then uses `vcSt`:
https://github.com/JacquesCarette/Drasil/blob/5dc10b54aefd5c6ddcef0ce15706b95760f6d7ca/code/drasil-lang/lib/Language/Drasil/Chunk/Quantity.hs#L71-L77
https://github.com/JacquesCarette/Drasil/blob/5dc10b54aefd5c6ddcef0ce15706b95760f6d7ca/code/drasil-lang/lib/Language/Drasil/Chunk/Quantity.hs#L111-L113

Meanwhile, `asVC'` uses `vc` directly:
https://github.com/JacquesCarette/Drasil/blob/5dc10b54aefd5c6ddcef0ce15706b95760f6d7ca/code/drasil-lang/lib/Language/Drasil/Chunk/Quantity.hs#L103-L105

The produced QuantityDicts are identical except for the Symbol appearance. `asVC` enforces that the symbol only shows itself for the "Implementation" stage, while `asVC'` allows the symbol to be shown for both "implementation" and "equational" stages.

*ASIDE*:

Naively, if we try to replace `asVC'`'s definition with just `asVC`, we get a runtime error:
e.g.,
```
-- | Helper for 'funcUID''.
asVC' :: Func -> QuantityDict
asVC' = asVC
```

results in:
```shell
glassbr: Prelude.foldl1: empty list
```

The error is a result of the lack of an Equational-level symbol for the `Func`s we use `asVC'` on. We can fix this by slightly altering `asVC` to also show the symbol at the "equational" time (e.g., replacing `implVar`'s `f Equational = Nothing` with `f Equational = sym` resolves this issue fairly quickly). This effectively removes `asVC'`, leaving only `asVC` (which also self-identifies as a hack).

Here, a larger, different, question occurs: Should QuantityDicts know about Equational vs Implementation stages? It looks like we might see a small benefit by replacing "QuantityDict" usage in `drasil-code` with a new "QuantityDict"-like container that's specifically intended for "implementation" only, and making the existing "QuantityDict" in `drasil-lang` "equational"-only. The benefit would strictly be in having separate representations of "symbol"s with the data around them tailored to either "code" or "equations".

**Comment by balacij** [comment]
Regarding the proposed solution we made, the general idea is to create a kind of QuantityDict that is specifically meant to be substituted by other things. The SRS will show it's intended usage and the name of the symbol (and potentially some basic description of what it _should_ be). The developer will then need to replace it accordingly in order for code to be generated and "usable" (upto the developer's definition being a "good" one).

4/4

**Comment by balacij** [comment]
A completely alternative solution is to _not_ treat this as a "choice", and to make data definitions allow for raw CodeExprs to be entered. In other words, instead of bringing the "Func" "up", we would create areas where "lower level pieces" can enter through data definitions. I'm not sure of how well this will work in terms of portability and reusability, but it could very well work for the existing languages.

**Comment by JacquesCarette** [comment]
Lots of stuff here!  Ok, let's be systematic:

> My thought was that we should use something slightly different from a QuantityDict because we would want Drasil to be able to automatically recognize which QuantityDicts need be replaced at "implementation time".

Not wrong - but over-design. QuantityDicts are abstract, full stop. It's the duty of the design to refine things, by implementing some bits, and not others. Some parts (like the interpolation function) are going to be needed to make things complete. The code generation will fail if it's not there. That's good, that's when the failure should happen. It should be possible to print out the SRS even without providing that implementation.

> The problem is that there are functions and data that is manually defined at the implementation level which are lacking appropriately abstracted placeholders. 

Almost correct. That is the problem that Brooks identified, but maybe the real problem is that there are functions and data that don't have an abstract (conceptual) definition. The concept of "an interpolation function" makes sense, and could be created in exactly that way. It's even generic. I think the data aspect will be example-specific, but also can exist as a 'concept'.

> Currently, to allow these manually written functions to "enter" the GlassBR program, they were forcibly "lifted" into it through asVC'

Right, and this is the "control inversion". It's the implementation pushing its way up into the spec. Makes complete sense given the development history, but is a hack (and could now be considered a bug, even). We need to invert this, by creating the concepts that represent these things as part of the SRS, and later in the flow provide an 'implementation'.

`asVC'` is the guilty party, which takes code and creates the `QuantityDict` **that should have existed all along**. What we instead need to do is to create that `QuantityDict` ourselves and, later, when `interpY` and `interpZ` are created, those `QuantityDict` should be used as part of the process.

> A completely alternative solution is to not treat this as a "choice", and to make data definitions allow for raw CodeExprs to be entered. In other words, instead of bringing the "Func" "up", we would create areas where "lower level pieces" can enter through data definitions. 

Much closer to what I had in mind. I don't necessarily want to do it via "data definitions" (that's still too much in SRS-land). These really belong in the abstract implementation stage. So it's fine to create `CodeExprs` by hand! That's one kind of refinement, it just happens late in the process. The point is that such `CodeExprs` should be actively associated with a specific `QuantityDict` (at the implementation phase). That's the piece of functionality we don't have.

**Comment by balacij** [comment]
>  Not wrong - but over-design. QuantityDicts are abstract, full stop. It's the duty of the design to refine things, by implementing some bits, and not others. Some parts (like the interpolation function) are going to be needed to make things complete. The code generation will fail if it's not there. That's good, that's when the failure should happen. It should be possible to print out the SRS even without providing that implementation.

I was thinking that `codeSpec` (code generator) would be the area where we ensure it's appropriately handled, so the SRS should still be able to be generated without an implementation (but this isn't very important...).  I can see how it's an over-design, however. Regarding the word "abstract", we maybe should have used a different name, but your point still stands, and makes sense to me. Thank you!

>  Almost correct. That is the problem that Brooks identified, but maybe the real problem is that there are functions and data that don't have an abstract (conceptual) definition. The concept of "an interpolation function" makes sense, and could be created in exactly that way. It's even generic. I think the data aspect will be example-specific, but also can exist as a 'concept'.

That makes sense, treating it as such doesn't sound all too different than what we were expecting if I'm understanding correctly.

>  asVC' is the guilty party, which takes code and creates the QuantityDict that should have existed all along. What we instead need to do is to create that QuantityDict ourselves and, later, when interpY and interpZ are created, those QuantityDict should be used as part of the process.

Yes, exactly.

> Much closer to what I had in mind. I don't necessarily want to do it via "data definitions" (that's still too much in SRS-land). These really belong in the abstract implementation stage. So it's fine to create CodeExprs by hand! That's one kind of refinement, it just happens late in the process. The point is that such CodeExprs should be actively associated with a specific QuantityDict (at the implementation phase). That's the piece of functionality we don't have.

Entering outside the SRS-land is probably the best then, I agree.

All in all, it sounds like @peter-michalski and I will need to take another look at the code again, but this sounds a bit simpler than what we were expecting (I think).

**Comment by balacij** [comment]
To avoid the complexity of ensuring code generates properly immediately, we're thinking of:
1. Get an unaltered version of GlassBR running locally
2. Turn off code generation in GlassBR (e.g., comment out the code generation code, but do not "make stabilize" this change, or commit it)
3. Create a replacement QuantityDicts for interpY and interpZ (hacky part: since `asVC` works by using the Func name as a faked UID, we need to make sure the UIDs for interpY and interpZ are "interpY" and "interpZ" respectively, and the terms and symbols alike)
4. Remove `asVC'`
5. Generate the SRS, and make sure it's still the same as before (e.g., stable)
6. Uncomment out the code generator, generate code, and ensure the results of the run code match with the originally run GlassBR.

Once we're here, we will want to figure out a way to tie the manually written `interpY` `Func` to the `interpY` `QuantityDict` properly. If we want to do this quickly, we can think of this as inverting the existing hack, and use the symbols of the `QuantityDict`s to replace the manually written Strings of the existing `Func`s written for interpY and interpZ.

**Comment by JacquesCarette** [comment]
Sorry, this fell off my radar. Sounds like a good plan.

**Comment by peter-michalski** [comment]
@balacij Based on our discussions, 9b159a3c8d651d207a4975d5894e950ea7290259 has the first attempt at implementing steps 1-4. Please let me know if anything stands out as incorrect. I will attempt to generate the SRS soon and will provide an update.

**Comment by balacij** [comment]
Looks great! :smile: 

To also get it written down somewhere too, we should also later look into what is needed for removing `asVC` (no `'`), they appear to be almost identical (upto Symbol depending on stage).

**Comment by peter-michalski** [comment]
@balacij The latest commit is in adadab34b17ab15f11e573c3327251eb47773bb6. Everything up to and including step 6 worked without error. The results of the run code match with the originally run GlassBR. We can move onto discussing how to tie the manually written `interpY` `Func` to the `interpY` `QuantityDict `

**Comment by peter-michalski** [comment]
@balacij Latest commit in 24459ca202ba784c6e51de9a4818731d5b9a3fd4. 
Created PR #2952 for this issue.

**Comment by balacij** [comment]
Looks good to me! :smile:

**Comment by peter-michalski** [comment]
This issue can be closed on account of PR #2952
